### DISCUSSION TITLE
Somehow (?) support `other-frame-prefix` and `other-window-prefix` with transient

### DISCUSSION THREAD

**Discussion by farblos** [magit/transient#305] [Support] [answered]
Lately, I got used to `C-x 5 5` and (less so) `C-x 4 4` to direct my buffers to new frames or windows, resp. So used that I tried `C-x 5 5 l l` from a magit buffer to get the logs shown in a new frame, which did not come out as expected.

Actually, in transient 20240201.1100 and Emacs 28.2 I get the messages:

```
Display next command buffer in a new frame...
Error while exiting transient: (error "Attempt to delete minibuffer or sole ordinary window")
```

after pressing `C-x 5 5 l` in a magit buffer.

I'm not sure whether this is technically doable and whether it fits into the overall tranient philosophy, but I think it would be nice if these prefixes would work for "final" transient commands as well.

Thanks!

**Comment by farblos** [comment]
Still getting the same behavior in transient 20240830.1817, FWIW.

**Comment by tarsius** [comment, marked as answer]
`l l` is actually two commands.

1. First `l` invokes `magit-log`, which displays a transient buffer and activates a transient keymap.
2. The second `l` invokes `magit-log-current`, because that is what it was bound to in the keymap, which was setup in step 1.

So to invoke `magit-log-current` with a prefix argument, you would have to type `l C-u l`.

Like `universal-argument`, `other-frame-prefix` affects the next command.

`l C-x 5 5 l` won't work out-of-the-box though. You first have to instruct Transient that it should be available in every transient menu and how it affects the "transient state":

```elisp
(keymap-set transient-map "M-5" #'other-frame-prefix)
(keymap-set transient-predicate-map "<other-frame-prefix>" #'transient--do-stay)
```

You could of course use `C-x 5 5`, instead of `M-5` as I have done here, but try that and you'll see that it is a bit inconvenient.

**Comment by farblos** [comment]
> `l l` is actually two commands.

I think sort of the point of this discussion is that I perceive these as an atomic key sequence, just like, say, `C-x r k`.

So I was hoping for some generic mechanism in transient that

- captures "all possible" Emacs prefix arguments (`universal-coding-system-argument` not having been mentioned yet) on transient prefix commands *without actually performing the effect of the prefix argument on the prefix command* and
- transfers the effect of the captured prefix argument to the suffix command.

But I had a look on how `other-frame-prefix` is implemented and decided that this would not be an easy thing to do.

So thanks for the pointers, I might give them a try instead, and feel free to close this discussion.

**Reply by tarsius** [reply]
That perception is inaccurate and now that this has been pointed out, you can adjust it.

Some prefix commands themselves accept universal arguments. That's one reason why they should not pass on their universal arguments to the subsequent suffix commands.

Note that the previous sentence ended with "suffix command**s**". That's another reason why universal arguments should be passed on from the prefix. If you used `C-u l - g l`, should the universal argument be passed along to `-g` or `l` or both? What if the universal argument has vastly different, possibly contradictory, meanings for the two (or more) involved commands?

Because one commonly invokes/"sets" infix arguments[^1] after invoking the prefix/menu command and before the final suffix command, you should avoid thinking of `<key for prefix> <key for suffix>` as an "atomic key sequence".

[^1]: Which are also implemented as commands.

**Reply by tarsius** [reply]
I'll play with `other-frame/window-prefix` and maybe add the above bindings by default.

(I would like to use the `C-x ...` bindings, but because that is currently used for some Transient-specific commands, which are displayed after typing the `C-x` prefix, this is not trivial, and will have to wait until I figure out how to deal with the mistake of using the `C-x` prefix for Transient-specific commands.)

> captures "all possible" Emacs prefix arguments 

What other such arguments exist?

**Comment by farblos** [comment]
> What other such arguments exist?

I use regularly and know about only those mentioned in this discussion already, which for completeness are:

- `universal-argument`
- `universal-coding-system-argument`
- `other-window-prefix`
- `other-frame-prefix`

Thanks for caring about this discussion, Transient, Magit, and in general!
